### PR TITLE
[WIP] Fix/lr schedulers update calling order

### LINKED ISSUE #7637: LR scheduler steps after saving checkpoint with iteration-based checkpointing
## 🐛 Bug

When checkpoints are saved using every_n_train_steps argument of ModelCheckpoint callback, checkpointing happens before lr_scheduler step being called. While it is the other way around with every_n_val_epochs argument.

In training_loop, on_train_batch_end (which saves checkpoint) is called before updating lr_scheduler. https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_loop.py#L500-L529

while on_evaluation_end is called after updating lrs. https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L996-L1011

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1bBkhGiKJoavp1O4Oi4kLWVnxohl5HR6M?usp=sharing

### Expected behavior

In the colab notebook, a training epoch is of 64 batches. When I save checkpoint using every_n_train_steps=64 or with every_n_val_epochs=1, the lr_scheduler step count should be the same. But the step count is 1 lesser with every_n_train_steps=64

### Environment

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### Additional context

cc @ananthsub

### PR DESCRIPTION
**Pull request by Lucklyric**
## What does this PR do?

Fixes #7637 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/checkpointing/test_trainer_checkpoint.py [hunk 1]
```diff
@@ -84,3 +87,40 @@ def validation_step(self, batch, batch_idx):
...
+    val_epoch_checkpoint_callback = ModelCheckpoint(dirpath=f"{tmpdir}/every_val_epoch", every_n_val_epochs=1)
+
+    model = TestModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=max_epochs,
+        callbacks=[train_step_checkpoint_callback, val_epoch_checkpoint_callback]
+    )
+    trainer.fit(model)
+    step_idx = data_length * max_epochs - 1
```
Comment (carmocca) [inline tests/checkpointing/test_trainer_checkpoint.py:119]: this could be just

`step = trainer.global_step - 1`

File: tests/checkpointing/test_trainer_checkpoint.py [hunk 2]
```diff
@@ -84,3 +87,40 @@ def validation_step(self, batch, batch_idx):
...
+            optimizer = torch.optim.SGD(self.parameters(), lr=0.001)
+            lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
+            lr_scheduler_dict = {'scheduler': lr_scheduler, 'interval': 'step'}
+            return [optimizer], [lr_scheduler_dict]
+
+        def train_dataloader(self):
+            return DataLoader(RandomDataset(32, data_length))
+
+    train_step_checkpoint_callback = ModelCheckpoint(dirpath=f"{tmpdir}/every_train_step", every_n_train_steps=1)
+    val_epoch_checkpoint_callback = ModelCheckpoint(dirpath=f"{tmpdir}/every_val_epoch", every_n_val_epochs=1)
```
Comment (carmocca) [inline tests/checkpointing/test_trainer_checkpoint.py:110]: Do we really need to save/load a checkpoint?

Can't you just test the state `on_train_batch_end`?

File: tests/checkpointing/test_trainer_checkpoint.py (hunk 1 above)
Comment (Lucklyric) [reply tests/checkpointing/test_trainer_checkpoint.py:119]: You are right!

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -494,6 +494,9 @@ def run_training_epoch(self):
             if batch_output.signal == -1:
                 break
 
+            # update LR schedulers
+            self.update_lr_schedulers('step')
```
[the remark this answers is not in the document]
Comment (Lucklyric) [reply pytorch_lightning/trainer/training_loop.py:498]: Dear @carmocca 
Exactly, thanks for pointing it out.
Will work around it further.

[the remark this answers is not in the document]
Comment (Lucklyric) [reply pytorch_lightning/trainer/training_loop.py:498]: Hi @carmocca , I realize that this could be a bit tricky. Because pl.module `on_train_batch_end` callbacks may log metrics, the LR scheduler must be updated following those callbacks. However, we must also make sure that a Checkpoint's `on_batch_batch_end` hook is called after the scheduler update.

[the remark this answers is not in the document]
Comment (Lucklyric) [reply pytorch_lightning/trainer/training_loop.py:498]: Hi @carmocca , in this case, the same issue may also exist for the validation part?
```python
        if not should_check_val or should_train_only:
            self.update_lr_schedulers('epoch')

        if should_train_only:
            self.check_checkpoint_callback(True)
```
if the metric is added in `on_validation_end` callbacks?

[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:498]: Long story short, in that snippet you only update the epoch schedulers if running only training
and we had another call to update epoch schedulers if running validation too.
So no issue

However, this was just updated in #7357. It has no conflicts with this PR but you should rebase master regardless

[the remark this answers is not in the document]
Comment (Lucklyric) [reply pytorch_lightning/trainer/training_loop.py:498]: Hi @carmocca , it seems that this issue has been fixed by your updates!
https://colab.research.google.com/drive/1bBkhGiKJoavp1O4Oi4kLWVnxohl5HR6M?usp=sharing

[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:498]: Oh! Sorry I didn't check myself.

Lovely when you unknowingly fix something :D
